csi/server/plugin/opensds/controller.go

```diff
@@ -91,8 +135,90 @@ func (p *Plugin) ControllerPublishVolume(
 	ctx context.Context,
 	req *csi.ControllerPublishVolumeRequest) (
 	*csi.ControllerPublishVolumeResponse, error) {
-	// TODO
-	return nil, nil
+
+	log.Println("start to ControllerPublishVolume")
+	defer log.Println("end to ControllerPublishVolume")
+
+	if support, errCode := p.CheckVersionSupport(req.Version); !support {
+		retry := false
+		if errCode == csi.Error_GeneralError_UNSUPPORTED_REQUEST_VERSION {
+			//if version not supported, caller should not try again.
+			retry = true
+		}
+		return newGeneralError(errCode, retry, "The request version is not supported."), nil
+	}
+
+	client := sdscontroller.GetClient("")
+
+	//check volume is exist
+	volSpec, errVol := client.GetVolume(req.VolumeHandle.Id)
+	if errVol != nil || volSpec == nil {
+		return newControllerPublishVolumeResponseError(csi.Error_ControllerPublishVolumeError_VOLUME_DOES_NOT_EXIST,
+			nil, "the volume is not exist."), nil
+	}
+
+	//need to check node exist?
+
+	attachments, err := client.ListVolumeAttachments()
+	if err != nil {
+		return newGeneralError(csi.Error_GeneralError_UNDEFINED, false, "Internal error."), nil
+	}
+
+	var attachNodes []*csi.NodeID
+	for _, attachSpec := range attachments {
+		if attachSpec.VolumeId == req.VolumeHandle.Id {
+			node := &csi.NodeID{
+				Values: map[string]string{
+					"ip": attachSpec.HostInfo.Ip,
+					"host": attachSpec.HostInfo.Host,
+				},
+			}
+			attachNodes = append(attachNodes, node)
+		}
+	}
+
+	if len(attachNodes) != 0 {
+		if len(attachNodes) >= VOLUME_ATTACH_NODES_MAX {
+			return newControllerPublishVolumeResponseError(csi.Error_ControllerPublishVolumeError_MAX_ATTACHED_NODES,
+				attachNodes, "the node to attath has reach max."), nil
+		}
+
+		//if the volume has been published, but without MULTI_NODE capability, return error.
+		mode := req.VolumeCapability.AccessMode.Mode
+		if mode != csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER &&
+			mode != csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY &&
+			mode != csi.VolumeCapability_AccessMode_MULTI_NODE_SINGLE_WRITER {
+			return newControllerPublishVolumeResponseError(csi.Error_ControllerPublishVolumeError_VOLUME_ALREADY_PUBLISHED,
+				attachNodes, "the volume has been published to another node."), nil
+		}
+	}
+
+	attachReq := &model.VolumeAttachmentSpec{
+		VolumeId: req.VolumeHandle.Id,
+		HostInfo: &model.HostInfo{
+			Ip: req.NodeId.Values["ip"],
+			Host: req.NodeId.Values["host"],
+		},
+	}
+	attachSpec, errAttach := client.CreateVolumeAttachment(client2.VolumeAttachmentBuilder(attachReq))

```

Comment:
Maybe you can use "client.CreateVolumeAttachment(attachReq)" directly.
So we can avoid to refer the "client2".

Reply:
ok, don't convert attachReq to VolumeAttachmentBuilder